fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,451 @@
+                                     "result of corresponding data instance.")
+
+                case.fail_id.append(data_id[result == False])
+
+    def on_epoch_end(self, data: Data) -> None:
+        for case in self.aggregate_cases:
+            result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            if not isinstance(result, (bool, np.bool_)):
+                raise TypeError(f"In test with description \'{case.description}\': "
+                                "Criteria return of aggregate-case test needs to be a bool.")
+            case.result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            case.input_val = {var_name: self._to_serializable(data[var_name]) for var_name in case.criteria_inputs}
+
+    def on_end(self, data: Data) -> None:
+        for case in self.instance_cases:
+            case_dict = {"test_type": "per-instance", "description": case.description}
+            result = np.hstack(case.result)
+            fail_num = np.sum(result == False)
+            case_dict["passed"] = self._to_serializable(fail_num <= case.fail_threshold)
+            case_dict["fail_threshold"] = case.fail_threshold
+            case_dict["fail_number"] = self._to_serializable(fail_num)
+            if self.data_id:
+                fail_id = np.hstack(case.fail_id)
+                case_dict["fail_id"] = self._to_serializable(fail_id)
+            self.json_summary["tests"].append(case_dict)
+
+        for case in self.aggregate_cases:
+            case_dict = {"test_type": "aggregate", "description": case.description}
+            case_dict["passed"] = self._to_serializable(case.result)
+            case_dict["inputs"] = case.input_val
+            self.json_summary["tests"].append(case_dict)
+
+        self.json_summary["execution_time(s)"] = time() - self.json_summary["execution_time(s)"]
+
+        self._dump_json()
+        self._init_document()
+        self._document_test_result()
+        self._dump_pdf()
+
+    def _initialize_json_summary(self) -> None:
+        """Initialize json summary
+        """
+        self.json_summary = {
+            "title": self.test_title, "timestamp": str(datetime.now()), "execution_time(s)": time(), "tests": []
+        }
+
+    def _sanitize_report_name(self) -> None:
+        """Sanitize report name and make it class attribute
+        """
+        exp_name = self.system.summary.name
+        if not exp_name:
+            raise RuntimeError("TestReport require an experiment name to be provided in estimator.fit()")
+        # Convert the experiment name to a report name (useful for saving multiple experiments into same directory)
+        report_name = "".join('_' if c == ' ' else c for c in exp_name
+                              if c.isalnum() or c in (' ', '_')).rstrip("_").lower()
+        report_name = re.sub('_{2,}', '_', report_name) + "_ModelEval"
+        self.report_name = report_name or '_ModelEval'
```

Comment:
no need for the 'or' here since you're already adding "_ModelEval" in the prior line